looks like a test needs to be adjusted
```
Error: core] [ERROR] Failures: 
Error: core] [ERROR]   XQueryContextTest.testReDeclareNamespaceEmptyPrefixFail:243 expected:<err:XQST00[33] Cannot bind prefix ...> but was:<err:XQST00[66] Cannot bind prefix ...>
[exist-core] [INFO] 
Error: core] [ERROR] Tests run: 6592, Failures: 1, Errors: 0, Skipped: 106
```

--- a/exist-core/src/main/antlr/org/exist/xquery/parser/XQueryTree.g
+++ b/exist-core/src/main/antlr/org/exist/xquery/parser/XQueryTree.g
@@ -387,7 +387,7 @@ throws PermissionDeniedException, EXistException, XPathException
             )
             {
                 if (orderempty)
-                    throw new XPathException(prolog_AST_in, ErrorCodes.XQST0065, "Ordering mode already declared.");
+                    throw new XPathException(prolog_AST_in, ErrorCodes.XQST0069, "Empty order declaration already declared.");
                 orderempty = true;
             }
         )
@@ -454,7 +454,7 @@ throws PermissionDeniedException, EXistException, XPathException
             {
                 // ignored
                 if (construction)
-                    throw new XPathException(prolog_AST_in, ErrorCodes.XQST0069, "Construction already declared.");
+                    throw new XPathException(prolog_AST_in, ErrorCodes.XQST0067, "Construction already declared.");
                 construction = true;
             }
         )
@@ -734,25 +734,15 @@ throws PermissionDeniedException, EXistException, XPathException
         targetURI:STRING_LITERAL
         ( uriList [uriList] )?
         {
-            if ("".equals(targetURI.getText()) && nsPrefix != null) {
-                    throw new XPathException(s, ErrorCodes.XQST0057, "A schema without target namespace (zero-length string target namespace) may not bind a namespace prefix: " + nsPrefix);
-            }
             if (nsPrefix != null) {
                 if (declaredNamespaces.get(nsPrefix) != null)
                     throw new XPathException(s, ErrorCodes.XQST0033, "Prolog contains " +
                                              "multiple declarations for namespace prefix: " + nsPrefix);
                 declaredNamespaces.put(nsPrefix, targetURI.getText());
             }
-            try {
-                context.declareNamespace(nsPrefix, targetURI.getText());
-                staticContext.declareNamespace(nsPrefix, targetURI.getText());
-                // We currently do nothing with eventual location hints. /ljo
-            } catch(XPathException xpe) {
-                xpe.prependMessage("err:XQST0059: Error found while loading schema " + nsPrefix + ": ");
-                throw xpe;
+            if (s != null) {
+                throw new XPathException(s, ErrorCodes.XQST0009, "The eXist-db XQuery implementation does not support the Schema Import Feature.");
             }
-            // We ought to do this for now until Dannes can say it works. /ljo
-            //throw new XPathException(s, ErrorCodes.XQST0009, "The eXist-db XQuery implementation does not support the Schema Import Feature quite yet.");
         }
     )
     ;

--- a/exist-core/src/main/java/org/exist/xquery/DynamicAttributeConstructor.java
+++ b/exist-core/src/main/java/org/exist/xquery/DynamicAttributeConstructor.java
@@ -109,12 +109,12 @@ public class DynamicAttributeConstructor extends NodeConstructor {
             	try {
             		qn = QName.parse(context, nameSeq.getStringValue(), null);
 		    	} catch (final QName.IllegalQNameException e) {
-					throw new XPathException(this, ErrorCodes.XPTY0004, "'" + nameSeq.getStringValue() + "' is not a valid attribute name");
+					throw new XPathException(this, ErrorCodes.XQDY0074, "'" + nameSeq.getStringValue() + "' is not a valid attribute name");
 				}
 
             //Not in the specs but... makes sense
             if(!XMLNames.isName(qn.getLocalPart()))
-            	{throw new XPathException(this, ErrorCodes.XPTY0004, "'" + qn.getLocalPart() + "' is not a valid attribute name");}
+            	{throw new XPathException(this, ErrorCodes.XQDY0074, "'" + qn.getLocalPart() + "' is not a valid attribute name");}
             
             if ("xmlns".equals(qn.getLocalPart()) && qn.getNamespaceURI().isEmpty())
             	{throw new XPathException(this, ErrorCodes.XQDY0044, "'" + qn.getLocalPart() + "' is not a valid attribute name");}

--- a/exist-core/src/main/java/org/exist/xquery/ElementConstructor.java
+++ b/exist-core/src/main/java/org/exist/xquery/ElementConstructor.java
@@ -226,7 +226,7 @@ public class ElementConstructor extends NodeConstructor {
                     try {
                         attrQName = QName.parse(context, constructor.getQName(), XMLConstants.NULL_NS_URI);
                     } catch (final QName.IllegalQNameException e) {
-                        throw new XPathException(this, ErrorCodes.XPTY0004, "'" + constructor.getQName() + "' is not a valid attribute name");
+                        throw new XPathException(this, ErrorCodes.XQDY0074, "'" + constructor.getQName() + "' is not a valid attribute name");
                     }
 
                     final String namespaceURI = attrQName.getNamespaceURI();
@@ -288,7 +288,7 @@ public class ElementConstructor extends NodeConstructor {
                     try {
                         qn = QName.parse(context, qnitem.getStringValue());
                     } catch (final QName.IllegalQNameException e) {
-                        throw new XPathException(this, ErrorCodes.XPTY0004, "'" + qnitem.getStringValue() + "' is not a valid element name");
+                        throw new XPathException(this, ErrorCodes.XQDY0074, "'" + qnitem.getStringValue() + "' is not a valid element name");
                     } catch (final XPathException e) {
                         e.setLocation(getLine(), getColumn(), getSource());
                         throw e;
@@ -306,7 +306,7 @@ public class ElementConstructor extends NodeConstructor {
 
             //Not in the specs but... makes sense
             if (!XMLNames.isName(qn.getLocalPart())) {
-                throw new XPathException(this, ErrorCodes.XPTY0004, "'" + qnitem.getStringValue() + "' is not a valid element name");
+                throw new XPathException(this, ErrorCodes.XQDY0074, "'" + qnitem.getStringValue() + "' is not a valid element name");
             }
 
             // add namespace declaration nodes

--- a/exist-core/src/main/java/org/exist/xquery/EnclosedExpr.java
+++ b/exist-core/src/main/java/org/exist/xquery/EnclosedExpr.java
@@ -190,6 +190,8 @@ public class EnclosedExpr extends PathExpr {
                         } catch (DOMException e) {
                             if (e.code == DOMException.NAMESPACE_ERR) {
                                 throw new XPathException(this, ErrorCodes.XQDY0102, e.getMessage());
+                            } else if (e.code == DOMException.INUSE_ATTRIBUTE_ERR) {
+                                throw new XPathException(this, ErrorCodes.XQDY0025, e.getMessage());
                             } else {
                                 throw new XPathException(this, e.getMessage(), e);
                             }

--- a/exist-core/src/main/java/org/exist/xquery/RangeSequence.java
+++ b/exist-core/src/main/java/org/exist/xquery/RangeSequence.java
@@ -196,13 +196,13 @@ public class RangeSequence extends AbstractSequence {
 
     @Override
     public NodeSet toNodeSet() throws XPathException {
-        throw new XPathException(this, "Type error: the sequence cannot be converted into" +
+        throw new XPathException(this, ErrorCodes.XPTY0019, "Type error: the sequence cannot be converted into" +
                 " a node set. Item type is xs:integer");
     }
 
     @Override
     public MemoryNodeSet toMemNodeSet() throws XPathException {
-        throw new XPathException(this, "Type error: the sequence cannot be converted into" +
+        throw new XPathException(this, ErrorCodes.XPTY0019, "Type error: the sequence cannot be converted into" +
                 " a memory node set. Item type is xs:integer");
     }
 

--- a/exist-core/src/main/java/org/exist/xquery/VariableImpl.java
+++ b/exist-core/src/main/java/org/exist/xquery/VariableImpl.java
@@ -258,7 +258,7 @@ public class VariableImpl implements Variable {
 				}
 			}
 
-        	throw new XPathException(getValue(),
+        	throw new XPathException(getValue(), ErrorCodes.XPTY0004,
 					Messages.getMessage(Error.VAR_TYPE_MISMATCH,
 							toString(),
 							type.toString(),

--- a/exist-core/src/main/java/org/exist/xquery/XQueryContext.java
+++ b/exist-core/src/main/java/org/exist/xquery/XQueryContext.java
@@ -875,7 +875,9 @@ public class XQueryContext implements BinaryValueManager, Context {
         }
         //Forbids rebinding the *same* prefix in a *different* namespace in this *same* context
         if (!nonNullUri.equals(prevURI)) {
-            throw new XPathException(rootExpression, ErrorCodes.XQST0033,
+            // XQST0066 for default element namespace redeclaration, XQST0033 for other prefixes
+            final ErrorCodes.ErrorCode errorCode = nonNullPrefix.isEmpty() ? ErrorCodes.XQST0066 : ErrorCodes.XQST0033;
+            throw new XPathException(rootExpression, errorCode,
                     "Cannot bind prefix '" + prefix + "' to '" + nonNullUri + "' it is already bound to '" + prevURI + "'");
         }
 

--- a/exist-core/src/main/java/org/exist/xquery/functions/fn/FunContainsToken.java
+++ b/exist-core/src/main/java/org/exist/xquery/functions/fn/FunContainsToken.java
@@ -90,7 +90,7 @@ public class FunContainsToken extends BasicFunction {
         Collator collator = context.getDefaultCollator();
 
         if (args.length > 2 && !args[2].isEmpty()) {
-            collator = context.getCollator(args[2].getStringValue());
+            collator = context.getCollator(args[2].getStringValue(), ErrorCodes.FOCH0002);
         }
 
         /* return true only if some fragment matches the trimmed token under current collation */

--- a/exist-core/src/main/java/org/exist/xquery/functions/fn/FunIndexOf.java
+++ b/exist-core/src/main/java/org/exist/xquery/functions/fn/FunIndexOf.java
@@ -28,6 +28,7 @@ import org.exist.xquery.Cardinality;
 import org.exist.xquery.Constants.Comparison;
 import org.exist.xquery.Constants.StringTruncationOperator;
 import org.exist.xquery.Dependency;
+import org.exist.xquery.ErrorCodes;
 import org.exist.xquery.Function;
 import org.exist.xquery.FunctionSignature;
 import org.exist.xquery.Profiler;
@@ -129,7 +130,7 @@ public class FunIndexOf extends BasicFunction {
     		Collator collator;
     		if (getSignature().getArgumentCount() == 3) {
     			final String collation = args[2].getStringValue();
-    			collator = context.getCollator(collation);
+    			collator = context.getCollator(collation, ErrorCodes.FOCH0002);
     		} else
     			{collator = context.getDefaultCollator();}
     		result = new ValueSequence();

--- a/exist-core/src/main/java/org/exist/xquery/functions/fn/FunSort.java
+++ b/exist-core/src/main/java/org/exist/xquery/functions/fn/FunSort.java
@@ -237,7 +237,7 @@ public class FunSort extends BasicFunction {
         return context.getDefaultCollator();
       }
       final String collationURI = args[pos].getStringValue();
-      return context.getCollator(collationURI);
+      return context.getCollator(collationURI, ErrorCodes.FOCH0002);
     } else {
       return context.getDefaultCollator();
     }

--- a/exist-core/src/main/java/org/exist/xquery/value/AnyURIValue.java
+++ b/exist-core/src/main/java/org/exist/xquery/value/AnyURIValue.java
@@ -136,7 +136,7 @@ public class    AnyURIValue extends AtomicValue {
             try {
                 XmldbURI.xmldbUriFor(escapedString);
             } catch (final URISyntaxException ex) {
-                throw new XPathException(getExpression(), 
+                throw new XPathException(getExpression(), ErrorCodes.XQST0046,
                         "Type error: the given string '" + s + "' cannot be cast to " + Type.getTypeName(getType()));
             }
         }

--- a/exist-core/src/main/java/org/exist/xquery/value/ArrayListValueSequence.java
+++ b/exist-core/src/main/java/org/exist/xquery/value/ArrayListValueSequence.java
@@ -40,6 +40,7 @@ import org.exist.dom.persistent.NewArrayNodeSet;
 import org.exist.dom.persistent.NodeProxy;
 import org.exist.dom.persistent.NodeSet;
 import org.exist.numbering.NodeId;
+import org.exist.xquery.ErrorCodes;
 import org.exist.xquery.Expression;
 import org.exist.xquery.NodeTest;
 import org.exist.xquery.XPathException;
@@ -403,7 +404,7 @@ public class ArrayListValueSequence extends AbstractSequence implements MemoryNo
 //            }
             return set;
         } else {
-            throw new XPathException((Expression) null, "Type error: the sequence cannot be converted into" +
+            throw new XPathException((Expression) null, ErrorCodes.XPTY0019, "Type error: the sequence cannot be converted into" +
                     " a node set. Item type is " + Type.getTypeName(itemType));
         }
     }
@@ -415,7 +416,7 @@ public class ArrayListValueSequence extends AbstractSequence implements MemoryNo
         }
 
         if (itemType == Type.ANY_TYPE || !Type.subTypeOf(itemType, Type.NODE)) {
-            throw new XPathException((Expression) null, "Type error: the sequence cannot be converted into" +
+            throw new XPathException((Expression) null, ErrorCodes.XPTY0019, "Type error: the sequence cannot be converted into" +
                     " a node set. Item type is " + Type.getTypeName(itemType));
         }
         for (final Item value : values) {

--- a/exist-core/src/main/java/org/exist/xquery/value/OrderedValueSequence.java
+++ b/exist-core/src/main/java/org/exist/xquery/value/OrderedValueSequence.java
@@ -310,7 +310,7 @@ public class OrderedValueSequence extends AbstractSequence {
             }
             return set;
         } else {
-            throw new XPathException((Expression) null, "Type error: the sequence cannot be converted into" +
+            throw new XPathException((Expression) null, ErrorCodes.XPTY0019, "Type error: the sequence cannot be converted into" +
                     " a node set. Item type is " + Type.getTypeName(itemType));
         }
     }
@@ -338,7 +338,7 @@ public class OrderedValueSequence extends AbstractSequence {
             return MemoryNodeSet.EMPTY;
         }
         if (itemType == Type.ANY_TYPE || !Type.subTypeOf(itemType, Type.NODE)) {
-            throw new XPathException((Expression) null, "Type error: the sequence cannot be converted into" +
+            throw new XPathException((Expression) null, ErrorCodes.XPTY0019, "Type error: the sequence cannot be converted into" +
                     " a node set. Item type is " + Type.getTypeName(itemType));
         }
         for (int i = 0; i < count; i++) {

--- a/exist-core/src/main/java/org/exist/xquery/value/StringValue.java
+++ b/exist-core/src/main/java/org/exist/xquery/value/StringValue.java
@@ -358,7 +358,7 @@ public class StringValue extends AtomicValue {
             }
             return charNumber;
         } catch (final NumberFormatException e) {
-            throw new XPathException(expression, "Unknown character reference: " + buf);
+            throw new XPathException(expression, ErrorCodes.XQST0090, "Unknown character reference: " + buf);
         }
     }
 

--- a/exist-core/src/main/java/org/exist/xquery/value/ValueSequence.java
+++ b/exist-core/src/main/java/org/exist/xquery/value/ValueSequence.java
@@ -305,7 +305,7 @@ public class ValueSequence extends AbstractSequence implements MemoryNodeSet {
             }
             return set;
         } else {
-            throw new XPathException((Expression) null, "Type error: the sequence cannot be converted into" +
+            throw new XPathException((Expression) null, ErrorCodes.XPTY0019, "Type error: the sequence cannot be converted into" +
                     " a node set. Item type is " + Type.getTypeName(itemType));
         }
     }
@@ -316,7 +316,7 @@ public class ValueSequence extends AbstractSequence implements MemoryNodeSet {
             return MemoryNodeSet.EMPTY;
         }
         if (itemType == Type.ANY_TYPE || !Type.subTypeOf(itemType, Type.NODE)) {
-            throw new XPathException((Expression) null, "Type error: the sequence cannot be converted into" +
+            throw new XPathException((Expression) null, ErrorCodes.XPTY0019, "Type error: the sequence cannot be converted into" +
                     " a node set. Item type is " + Type.getTypeName(itemType));
         }
         for (int i = 0; i <= size; i++) {

--- a/exist-core/src/test/java/org/exist/xquery/XQueryContextTest.java
+++ b/exist-core/src/test/java/org/exist/xquery/XQueryContextTest.java
@@ -240,7 +240,7 @@ public class XQueryContextTest {
             context.declareNamespace("", "new-default");
             fail("empty prefix was rebound");
         } catch (XPathException e) {
-            assertEquals("err:XQST0033 Cannot bind prefix '' to 'new-default' it is already bound to 'default'",
+            assertEquals("err:XQST0066 Cannot bind prefix '' to 'new-default' it is already bound to 'default'",
                     e.getMessage());
             assertEquals("default",  context.staticNamespaces.get(""));
         }

